ocaml/lambda/translcore.ml

```diff
@@ -1599,7 +1632,11 @@ and transl_match ~scopes ~arg_sort ~return_sort ~return_type ~loc ~env
         in
         val_cases, exn_case :: exn_cases, static_handlers
     | Some pv, Some pe ->
-        assert (c_guard = None);
+        let rhs_exp =
+          match c_rhs with
+          | Simple_rhs rhs -> rhs
+          | Boolean_guarded_rhs _ | Pattern_guarded_rhs _ -> assert false
```

Comment:
The `assert false` was never tripped before because:

```ocaml
# exception Exn;;
# let f x = match x with | Some x | exception x when x = Exn -> ();;
Error: Mixing value and exception patterns under when-guards is not supported.
```
Is this also true for when pattern guards?

  * If not, then the `assert false` could be tripped and we should fix that.
  * If so, we should write a test that evidences that.

Reply:
It is true, added such a test.